src/Http/Pool.php

```diff
@@ -47,71 +50,77 @@ class Pool implements PoolContract
      *
      * How many requests will be sent at once.
      *
-     * @var \Closure|int
+     * @var int|\Closure(int $pendingRequests): int
      */
-    protected Closure|int $concurrency;
+    protected int|Closure $concurrency;
 
     /**
      * Constructor
      *
      * @param \Saloon\Http\Connector $connector
-     * @param iterable<\GuzzleHttp\Promise\PromiseInterface|\Saloon\Contracts\Request>|callable $requests
-     * @param int|callable $concurrency
-     * @param callable|null $responseHandler
-     * @param callable|null $exceptionHandler
+     * @param iterable<\GuzzleHttp\Promise\PromiseInterface|\Saloon\Contracts\Request>|callable(\Saloon\Contracts\Connector): iterable<\GuzzleHttp\Promise\PromiseInterface|\Saloon\Contracts\Request> $requests
+     * @param int|callable(int $pendingRequests): (int) $concurrency
+     * @param callable(\Saloon\Contracts\Response, array-key $key, \GuzzleHttp\Promise\PromiseInterface $poolAggregate): (void)|null $responseHandler
+     * @param callable(mixed $reason, array-key $key, \GuzzleHttp\Promise\PromiseInterface $poolAggregate): (void)|null $exceptionHandler
      */
     public function __construct(Connector $connector, iterable|callable $requests = [], int|callable $concurrency = 5, callable|null $responseHandler = null, callable|null $exceptionHandler = null)
     {
         $this->connector = $connector;
         $this->setRequests($requests);
-        $this->concurrency = $concurrency;
-        $this->responseHandler = $responseHandler;
-        $this->exceptionHandler = $exceptionHandler;
+        $this->setConcurrency($concurrency);
+
+        if (! is_null($responseHandler)) {
+            $this->withResponseHandler($responseHandler);
+        }
+
+        if (! is_null($exceptionHandler)) {
+            $this->withExceptionHandler($exceptionHandler);
+        }
     }
 
     /**
      * Specify a callback to happen for each successful request
      *
-     * @param callable $callable
+     * @param callable(\Saloon\Contracts\Response, array-key $key, \GuzzleHttp\Promise\PromiseInterface $poolAggregate): (void) $callable
      * @return $this
      */
     public function withResponseHandler(callable $callable): static
     {
-        $this->responseHandler = $callable;
+        $this->responseHandler = $callable(...);
```

Comment:
What does the ellipses do here?

Reply:
It's the first-class callable syntax.
They turn any callable into a Closure, in the same fashion as as Closure::fromCallable().
It actually worked without, but is technically wrong as a Closure is an invokable class (callable), but a callable isn't necessarily a Closure.
So changed it to explicitly turn them into Closures, to fit the type hint of the property.

https://www.php.net/manual/en/functions.first_class_callable_syntax.php

Reply:
You learn things every day, thanks Juse!